File: src/main/java/org/project/model/response/ProductResponse.java
package org.project.model.response;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.project.enums.Label;
import org.project.enums.ProductStatus;
import org.project.enums.ProductType;

import java.math.BigDecimal;
import java.util.List;

@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class ProductResponse {
    private Long id;
    private String name;
    private String description;
    private BigDecimal price;
    private String unit;
    private Integer stockQuantity;
    private String imageUrl;
    private ProductType productType;
    private ProductStatus productStatus;
    private Label label;
    private Double averageRating;
    private Long reviewCount;
    private List<ProductAdditionalInfoResponse> productAdditionalInfoEntities;
}

File: src/main/resources/templates/frontend/assets/js/model/response/ProductResponse.js
import {CategoryResponse} from "/templates/frontend/assets/js/model/response/CategoryResponse.js";
import {
    ProductAdditionalInfoResponse
} from "/templates/frontend/assets/js/model/response/ProductAdditionalInfoResponse.js";

export class ProductResponse {
    constructor(id, name, description, imageUrl, price, categoryEntities, productAdditionalInfoEntities, reviewCount, averageRating) {
        this.id = id;
        this.name = name;
        this.description = description;
        this.imageUrl = imageUrl;
        this.price = price;
        this.categoryEntities = categoryEntities;
        this.productAdditionalInfoEntities = productAdditionalInfoEntities;
        this.reviewCount = reviewCount;
        this.averageRating = averageRating;
    }

    static fromJson(json) {
        const productResponse = new ProductResponse(
            json.id,
            json.name,
            json.description,
            json.imageUrl,
            json.price,
            json.categoryEntities,
            json.productAdditionalInfoEntities,
            json.reviewCount,
            json.averageRating
        );

        if (json.categoryEntities) {
            productResponse.categoryEntities = CategoryResponse.fromJsonArray(json.categoryEntities);
        }
        if (json.productAdditionalInfoEntities) {
            productResponse.productAdditionalInfoEntities = ProductAdditionalInfoResponse.fromJsonArray(json.productAdditionalInfoEntities);
        }
        return productResponse;
    }

    static fromJsonArray(jsonArray) {
        return jsonArray.map(json => ProductResponse.fromJson(json));
    }
}
